File: 54.py
class Solution:
    # @param {integer[][]} matrix
    # @return {integer[]}
    def spiralOrder(self, matrix):
        if matrix == []:
            return matrix
        m = len(matrix)
        if m == 1:
            return matrix[0]
        if matrix[0] == []:
            return []
            
        n = len(matrix[0])
        if n == 1:
            return [matrix[i][0] for i in range(m)]
        
        spiral_order = []
        
        spiral_order.extend(matrix[0])
        for k in range(1, m-1):
            spiral_order.append(matrix[k][n-1])
            matrix[k][n-1:] = []
        
        spiral_order.extend(matrix[m-1][::-1])
        
        for k in range(m-2,0,-1):
            spiral_order.append(matrix[k][0])
            matrix[k][0:1] = []
        
        sub_order = self.spiralOrder(matrix[1:m-1])
        spiral_order.extend(sub_order)
        
        return spiral_order
        
        
        

File: 54.cpp
#include <vector>
#include <iostream>
using namespace std;



class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> result;
        int m = matrix.size();
        int n = matrix[0].size();
        if( m == 0 || n == 0)
            return result;
            
        int count = 1;
        int i = 0 ;
        int j = 0 ;
        int row = 0;
        int span = n;
        while (count <= m*n)
        {
            i = row;
            j = row;
            
            // left to right
            for(;j<=row+span-1;j++)
            {
 
                result.push_back(matrix[i][j]);
                count++;
            }
            j--;
            
            // down 

            for(++i;i<=row+span-1;i++)
            {
                result.push_back(matrix[i][j]);
                count++;
            }
            i--;
            
            // right to left
            for(--j;j>=row;j--)
            {
                result.push_back(matrix[i][j]);
                count++;
            }
            j++;
            
            // up
            for(--i;i>=row+1;i--)
            {
                result.push_back(matrix[i][j]);
                count++;
            }
            
            row++;
            span -= 2;
            
        }
        
        return result;
        
        
    }
};


int main()
{
    Solution sol = new Solution();
    vector<vector<int>> matrix;
    matrix = {};
    cout<< sol.spiralOrder(matrix)<<endl;
    return 0;
}